private/pkg/dependenciestree/dependenciestree.go

```diff
@@ -0,0 +1,213 @@
+// - c: [d]
+// - d: [b] // circular dependency: b.c.d.b
+
+package dependenciestree
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+)
+
+// DependenciesTree is a tree that holds nodes depending on other nodes. One level deep.
+type DependenciesTree interface {
+	// NewRootNode adds a new node at the root of the tree, with its dependencies. Both root node name
+	// and dependencies names need to have values. Adding root nodes with invalid dependencies
+	// (external or circular) is possible, tree should be checked against the Validate func.
+	NewRootNode(rootNode string, dependencies map[string]struct{}) error
+	// Validate checks that the tree nodes have valid dependencies, with no circular dependencies.
+	Validate() error
+	// Sort validates that the tree is valid, and then returns the tree in topological order, leafs
+	// first. Useful for installation or update order.
+	Sort() ([]string, error)
+}
+
+// New initializes a new dependencies tree, ready to have root nodes added to it.
+func New(opts ...NewDependenciesTreeOption) DependenciesTree {
+	var config newDependenciesTreeOptions
+	for _, opt := range opts {
+		opt(&config)
+	}
+	return &depsTree{
+		tree:                      make(map[string]map[string]struct{}),
+		allowExternalDependencies: config.allowExternalDependencies,
+	}
+}
+
+// NewDependenciesTreeOption are options that can be passed when instantiating a new dependencies tree.
+type NewDependenciesTreeOption func(*newDependenciesTreeOptions)
+
+// NewDependenciesTreeWithAllowExternalDeps allows a root node in the tree to depend on external
+// dependencies, not present in other root nodes.
+func NewDependenciesTreeWithAllowExternalDeps() NewDependenciesTreeOption {
+	return func(opts *newDependenciesTreeOptions) {
+		opts.allowExternalDependencies = true
+	}
+}
+
+func (t *depsTree) NewRootNode(rootNode string, dependencies map[string]struct{}) error {
+	if rootNode == "" {
+		return errors.New("empty root node")
+	}
+	for dep := range dependencies {
+		if dep == "" {
+			return errors.New("empty dependency node")
+		}
+	}
+	t.tree[rootNode] = dependencies
+	return nil
+}
+
+func (t *depsTree) Validate() error {
+	for parent := range t.tree {
+		if err := t.validateDependencies(parent, []string{parent}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *depsTree) Sort() ([]string, error) {
+	if len(t.tree) == 0 {
+		return nil, nil
+	}
+	if err := t.Validate(); err != nil {
+		return nil, fmt.Errorf("cannot sort, invalid tree: %w", err)
+	}
+	pendingRootNodes := make(map[string]struct{}, len(t.tree))
+	for node := range t.tree {
+		pendingRootNodes[node] = struct{}{}
+	}
+	return t.sort(pendingRootNodes, nil)
+}
+
+type newDependenciesTreeOptions struct {
+	allowExternalDependencies bool
+}
+
+type depsTree struct {
+	tree                      map[string]map[string]struct{}
+	allowExternalDependencies bool
+}
+
+func (t *depsTree) validateDependencies(
```

Comment:
This package switches between `deps` and `dependencies` a lot, see above comments

Reply:
I'll default to `deps`.